pkg/envoy/envoy.go

```diff
@@ -152,115 +152,121 @@ func StartEnvoy(runDir, logPath string, baseID uint64) *Envoy {
-			select {
-			case started <- true:
-			default:
-			}
 
-			log.Infof("Envoy: Proxy started with pid %d", cmd.Process.Pid)
-			metrics.SubprocessStart.WithLabelValues(ciliumEnvoy).Inc()
-
-			// We do not return after a successful start, but watch the Envoy process
-			// and restart it if it crashes.
-			// Waiting for the process execution is done in the goroutime.
-			// The purpose of the "crash channel" is to inform the loop about their
-			// Envoy process crash - after closing that channel by the goroutime,
-			// the loop continues, the channel is recreated and the new process
-			// is watched again.
-			crashCh := make(chan struct{})
-			go func() {
-				if err := cmd.Wait(); err != nil {
-					log.WithError(err).Warn("Envoy: Proxy crashed")
-					// Avoid busy loop & hogging CPU resources by waiting before restarting envoy.
-					time.Sleep(100 * time.Millisecond)
-				}
-				close(crashCh)
-			}()
+				log.Infof("Envoy: Proxy started with pid %d", cmd.Process.Pid)
+				metrics.SubprocessStart.WithLabelValues(ciliumEnvoy).Inc()
+
+				// We do not return after a successful start, but watch the Envoy process
+				// and restart it if it crashes.
+				// Waiting for the process execution is done in the goroutime.
+				// The purpose of the "crash channel" is to inform the loop about their
+				// Envoy process crash - after closing that channel by the goroutime,
+				// the loop continues, the channel is recreated and the new process
+				// is watched again.
+				crashCh := make(chan struct{})
+				go func() {
+					if err := cmd.Wait(); err != nil {
+						log.WithError(err).Warn("Envoy: Proxy crashed")
+						// Avoid busy loop & hogging CPU resources by waiting before restarting envoy.
+						time.Sleep(100 * time.Millisecond)
+					}
+					close(crashCh)
+				}()
 
-			select {
-			case <-crashCh:
-				// Start Envoy again
-				continue
-			case <-e.stopCh:
-				log.Infof("Envoy: Stopping proxy with pid %d", cmd.Process.Pid)
-				if err := e.admin.quit(); err != nil {
-					log.WithError(err).Fatalf("Envoy: Envoy admin quit failed, killing process with pid %d", cmd.Process.Pid)
-
-					if err := cmd.Process.Kill(); err != nil {
-						log.WithError(err).Fatal("Envoy: Stopping Envoy failed")
-						e.errCh <- err
+				select {
+				case <-crashCh:
+					// Start Envoy again
+					continue
+				case <-e.stopCh:
+					log.Infof("Envoy: Stopping proxy with pid %d", cmd.Process.Pid)
+					if err := e.admin.quit(); err != nil {
+						log.WithError(err).Fatalf("Envoy: Envoy admin quit failed, killing process with pid %d", cmd.Process.Pid)
+
+						if err := cmd.Process.Kill(); err != nil {
+							log.WithError(err).Fatal("Envoy: Stopping Envoy failed")
+							e.errCh <- err
+						}
 					}
+					close(e.errCh)
+					return
 				}
-				close(e.errCh)
-				return
 			}
+		}()
+
+		if <-started {
+			return e
 		}
-	}()
 
-	if <-started {
+		return nil
+	} else {
+		// Case Envoy as DaemonSet
 		return e
 	}
```

Comment:
Small nit but if we did if externalEnvoyProxy { return e } in the beginning we could have avoided such a big diff.

Reply:
@aanm 

good point - thanks. i refactored it on this commit to ease reviewing for upcoming reviewers -  even though that part got refactored and moved anyway in the context of the whole PR.